llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1686,6 +1707,21 @@ Instruction *InstCombinerImpl::visitAShr(BinaryOperator &I) {
       if (match(Op0, m_OneUse(m_NSWSub(m_Value(X), m_Value(Y)))))
         return new SExtInst(Builder.CreateICmpSLT(X, Y), Ty);
     }
+
+    const APInt *MulC;
+    if (match(Op0, m_OneUse(m_NSWMul(m_Value(X), m_APInt(MulC)))) &&
+        (BitWidth > 2 && (*MulC - 1).isPowerOf2() &&
+         MulC->logBase2() == ShAmt &&
+         (ShAmt < BitWidth - 1))) /* Minus 1 for the sign bit */ {
+
+      // ashr (mul nsw (X, 2^N + 1)), N -> add nsw (X, ashr(X, N))
+      auto *NewAdd = BinaryOperator::CreateNSWAdd(
+          X,
+          Builder.CreateAShr(X, ConstantInt::get(Ty, ShAmt), "", I.isExact()));
+      NewAdd->setHasNoUnsignedWrap(
+          cast<OverflowingBinaryOperator>(Op0)->hasNoUnsignedWrap());
```

Comment:
Your proofs do not seem to cover all the relevant flag combinations, and indeed, preserving nuw does not appear to be valid if the ashr is not exact: https://alive2.llvm.org/ce/z/5vVrZw

Reply:
will fix now.

Reply:
Wait no, that proof uses an incorrect `ule` instead of `ult` precondition. Which I copied from your second alive2 link. ***WHICH DOES NOT VERIFY***. Can you please at least make sure that the proofs you submit actually verify?!

Reply:
https://alive2.llvm.org/ce/z/vyKvde, which matches the code

Reply:
https://alive2.llvm.org/ce/z/dRFsfV